File: prototype/rails/src/app/models/customer.rb
class Customer < ApplicationRecord
  prefixed_id :cus

  include EmailAddress
  include Messaging

  # The rows that move with a customer when an anonymous identity is absorbed
  # into a verified one, each by re-pointing the column it hangs from.
  # Conversations move through `Conversation#move_to`, since two threads of the
  # same shape have to fold into one.
  MERGED_ASSOCIATIONS = %i[
    favorites carts orders listing_events notifications sent_messages
  ].freeze

  # The standings the admin directory narrows the table to. `all` is the
  # filter a page carries when nobody has chosen one.
  STANDINGS = %w[all verified anonymous blocked].freeze

  # One line of the customers directory: a customer beside the counts the
  # table shows for them.
  Row = Data.define(:customer, :order_count, :favorite_count, :cart_line_count) do
    delegate :id, :email, :display_name, :anonymous?, :blocked?, to: :customer
  end

  # A merge from this customer's side: the visitor folded into them, or the
  # account they were folded into.
  Merge = Data.define(:direction, :other, :created_at)

  has_many :merges_absorbed, class_name: "CustomerMerge", dependent: :destroy, inverse_of: :customer
  has_one :merge_into, class_name: "CustomerMerge", foreign_key: :anonymous_customer_id,
    inverse_of: :anonymous_customer
  has_many :carts, dependent: :destroy
  has_many :favorites, dependent: :destroy
  has_many :listing_events, dependent: :nullify
  has_many :notifications, as: :recipient, dependent: :destroy
  has_many :orders, dependent: :restrict_with_error

  scope :verified, -> { where.not(email: nil) }
  scope :anonymous, -> { where(email: nil) }
  # An admin's block takes shopping and messaging away from a customer.
  # Nothing blocks a customer, so nobody stands blocked.
  scope :blocked, -> { none }
  scope :standing, ->(standing) {
    case standing
    when "verified" then verified
    when "anonymous" then anonymous
    when "blocked" then blocked
    end
  }

  # Every customer in the current scope with the counts the directory shows.
  # Each count is one grouped read for the whole table.
  def self.directory
    order_counts = Order.group(:customer_id).count
    favorite_counts = Favorite.group(:customer_id).count
    cart_line_counts = CartItem.joins(:cart).group("carts.customer_id").count

    order(:created_at, :id).map do |customer|
      Row.new(
        customer: customer,
        order_count: order_counts.fetch(customer.id, 0),
        favorite_count: favorite_counts.fetch(customer.id, 0),
        cart_line_count: cart_line_counts.fetch(customer.id, 0)
      )
    end
  end

  # The customer that owns the address once a link for it is followed:
  # a new account, the account already holding it, the anonymous row the
  # identity cookie points at, or the account that row is absorbed into.
  def self.claim(email, current: nil)
    anonymous = current if current&.anonymous?
    owner = find_by(email: email)

    return create!(email: email, email_verified_at: Time.current) if owner.nil? && anonymous.nil?
    return anonymous.claim_address(email) if owner.nil?

    owner.verify!
    anonymous ? owner.absorb(anonymous) : owner
  end

  # Returns nil when the cookie is absent, carries anything but a customer id,
  # or points at a customer that no longer exists.
  def self.from_cookie(value)
    id = PrefixedUlid.parse(value, :cus)
    return nil if id.nil?

    find_by(id: CustomerMerge.where(anonymous_customer_id: id).pick(:customer_id) || id)
  end

  def anonymous?
    email.nil?
  end

  # Whether an admin has taken shopping and messaging away from this customer.
  def blocked?
    blocks.any?
  end

  # The blocks an admin has placed on this customer, newest first. Nothing
  # blocks a customer, so the history is empty.
  def blocks
    []
  end

  # Every merge this customer was named in, oldest first, whichever side of it
  # they were on.
  def merges
    history = merges_absorbed.includes(:anonymous_customer).map do |merge|
      Merge.new(direction: "absorbed", other: merge.anonymous_customer, created_at: merge.created_at)
    end
    if merge_into
      history << Merge.new(direction: "folded_into", other: merge_into.customer, created_at: merge_into.created_at)
    end

    history.sort_by(&:created_at)
  end

  # A customer gives an address before they give a name, and a visitor gives
  # neither, so the display falls back through what is there.
  def display_name
    name.to_s.strip.presence || email.to_s.split("@").first.presence || "Visitor ##{id}"
  end

  # A merge hands the verified customer whatever cart the anonymous visitor was
  # filling, so one customer can own two. The one holding the most items is the
  # one they were shopping with.
  def current_cart
    carts.includes(:items).max_by { |cart| [ cart.items.size, cart.id ] } || carts.create!
  end

  # One button favorites and unfavorites, so what it does follows from what the
  # visitor has saved already. Returns :added or :removed.
  def toggle_favorite(listing, at: Time.current)
    saved = favorites.find_by(listing: listing)

    if saved
      saved.destroy!
      listing.record_event!("unfavorite", customer_id: id, at: at)

      return :removed
    end

    favorites.create!(listing: listing)
    listing.record_event!("favorite", customer_id: id, at: at)

    :added
  end

  def favorited?(listing)
    favorites.exists?(listing: listing)
  end

  def claim_address(email)
    update!(email: email, email_verified_at: Time.current)

    self
  end

  # A guest checkout can leave an address on a customer without verifying it;
  # clicking a link for that address settles it.
  def verify!
    update!(email_verified_at: Time.current) if email_verified_at.nil?

    self
  end

  # Takes over the history of an anonymous customer and returns self. The
  # anonymous row survives so a cookie still holding its id resolves forward
  # instead of starting the visitor over.
  def absorb(anonymous)
    Story.tell("customer.merge", "folding an anonymous visitor into the account",
      customer_id: id, anonymous_customer_id: anonymous.id) do |story|
      fold(anonymous)

      story.did("folded the anonymous visitor into the account",
        customer_id: id, anonymous_customer_id: anonymous.id)

      self
    end
  end

  private

  def fold(anonymous)
    transaction do
      MERGED_ASSOCIATIONS.each do |association|
        # Each association names the column it points back through: notifications
        # arrive at a polymorphic `recipient_id`, the rest at `customer_id`.
        foreign_key = self.class.reflect_on_association(association).foreign_key
        anonymous.public_send(association).update_all(foreign_key => id)
      end

      anonymous.conversations.each { |conversation| conversation.move_to(self) }
      merges_absorbed.create!(anonymous_customer: anonymous)
    end
  end
end

File: app/src/app/Models/Customer.php
<?php

declare(strict_types=1);

namespace App\Models;

use App\Domain\Customers\StandingFilter;
use App\Domain\Messaging\ParticipantName;
use App\Models\Concerns\HasPrefixedUlid;
use Database\Factories\CustomerFactory;
use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Attributes\Hidden;
use Illuminate\Database\Eloquent\Attributes\Scope;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\HasManyThrough;
use Illuminate\Database\Eloquent\Relations\HasOne;
use Illuminate\Database\Eloquent\Relations\MorphMany;
use Illuminate\Database\Eloquent\Relations\Relation;
use Illuminate\Foundation\Auth\User as Authenticatable;
use Illuminate\Notifications\Notifiable;
use LogicException;
use Override;

#[Fillable(['email', 'name', 'email_verified_at'])]
#[Hidden(['remember_token'])]
class Customer extends Authenticatable
{
    /** @use HasFactory<CustomerFactory> */
    use HasFactory;

    use HasPrefixedUlid;
    use Notifiable;

    public static function idPrefix(): string
    {
        return 'cus';
    }

    /**
     * @return array<string, string>
     */
    #[Override]
    protected function casts(): array
    {
        return [
            'email_verified_at' => 'datetime',
        ];
    }

    /** @return HasMany<Order, $this> */
    public function orders(): HasMany
    {
        return $this->hasMany(Order::class);
    }

    /** @return HasMany<Cart, $this> */
    public function carts(): HasMany
    {
        return $this->hasMany(Cart::class);
    }

    /** @return HasMany<Favorite, $this> */
    public function favorites(): HasMany
    {
        return $this->hasMany(Favorite::class);
    }

    /**
     * Every line in every cart the customer holds, so a page counts or lists
     * them without walking the carts first.
     *
     * @return HasManyThrough<CartItem, Cart, $this>
     */
    public function cartItems(): HasManyThrough
    {
        return $this->hasManyThrough(CartItem::class, Cart::class);
    }

    /**
     * Merges this customer absorbed: an anonymous visitor's rows folded into
     * this account when they verified an address.
     *
     * @return HasMany<CustomerMerge, $this>
     */
    public function mergesAsCustomer(): HasMany
    {
        return $this->hasMany(CustomerMerge::class);
    }

    /**
     * The merge that folded this row into someone else, which only an
     * anonymous customer has.
     *
     * @return HasMany<CustomerMerge, $this>
     */
    public function mergesAsAnonymous(): HasMany
    {
        return $this->hasMany(CustomerMerge::class, 'anonymous_customer_id');
    }

    /** @return BelongsToMany<Listing, $this> */
    public function favoriteListings(): BelongsToMany
    {
        return $this->belongsToMany(Listing::class, 'favorites');
    }

    /** @return HasMany<CustomerBlock, $this> */
    public function blocks(): HasMany
    {
        return $this->hasMany(CustomerBlock::class);
    }

    /** @return HasMany<Conversation, $this> */
    public function conversations(): HasMany
    {
        return $this->hasMany(Conversation::class);
    }

    /** @return MorphMany<Message, $this> */
    public function sentMessages(): MorphMany
    {
        return $this->morphMany(Message::class, 'sender');
    }

    /** @return HasOne<CustomerBlock, $this> */
    public function activeBlock(): HasOne
    {
        return $this->blocks()->one()->whereNull('lifted_at')->latestOfMany('created_at');
    }

    /**
     * Queries `activeBlock` fresh on every call, so a caller that never
     * eager-loaded it still gets an answer under strict mode.
     */
    public function currentBlock(): ?CustomerBlock
    {
        return $this->activeBlock()->first();
    }

    /**
     * A blocked customer keeps browsing, searching, and favoriting — this is
     * the one predicate the paths that buy something and the paths that post
     * a message both read.
     */
    public function canShop(): bool
    {
        return $this->currentBlock() === null;
    }

    /**
     * The admin's own words for why this customer cannot buy right now, for
     * the refusal that names them. Null when the customer is not blocked.
     */
    public function blockReason(): ?string
    {
        return $this->currentBlock()?->reason;
    }

    /**
     * A customer holds at most one cart. `MergeAnonymousCustomer` merges an
     * anonymous visitor's cart into the verified customer's own and retires
     * the anonymous cart, so there is never a second one to choose between.
     *
     * An unsaved row has no id a cart could be keyed by, so this refuses
     * rather than minting one for a visitor nothing has committed yet — a
     * caller reaching this on an unsaved row is the bug, not the row.
     * `cartOrNull()` is how a read tolerates one.
     */
    public function cart(): Cart
    {
        if (! $this->exists) {
            throw new LogicException('An unsaved customer has no cart to hold.');
        }

        return $this->carts()->first() ?? $this->carts()->create();
    }

    /**
     * The cart page's read: the customer's cart when one exists, and no
     * query at all — instead of a create — for a visitor who has never
     * added anything, saved or not.
     */
    public function cartOrNull(): ?Cart
    {
        return $this->exists ? $this->carts()->first() : null;
    }

    /**
     * A storefront visitor's row, once one exists, holds no address until they
     * give one; that row is theirs to claim on the day they verify one.
     */
    public function isAnonymous(): bool
    {
        return $this->email === null;
    }

    /**
     * A verified address is the one a card and an order history can hang on.
     */
    public function isVerified(): bool
    {
        return $this->email_verified_at !== null;
    }

    /**
     * What the admin console calls this customer: their name, the address
     * they verified, or — for a visitor who has given neither — their id.
     */
    public function displayName(): string
    {
        return $this->name ?? $this->email ?? $this->id;
    }

    /**
     * Everything the admin console's customer page reads off this row. Two
     * routes render that page — the page itself, and the message form on it
     * handing the page back when the message-post budget is spent — so the
     * eager loads it needs are named here once.
     */
    public function loadForConsole(): static
    {
        return $this->load([
            'activeBlock',
            'orders' => fn (Relation $orders) => $orders->withCount('items')->orderByDesc('placed_at')->orderByDesc('id'),
            'blocks' => fn (Relation $blocks) => $blocks->orderByDesc('created_at')->orderByDesc('id'),
            'favorites.listing',
            'cartItems.listing',
            'mergesAsCustomer',
            'mergesAsAnonymous',
        ]);
    }

    /**
     * Takes the rows a moderation decision is judged against for update. A
     * block is refused when one already stands, and that check queries the
     * `customer_blocks` table alone; locking this row does not guard it, so
     * both admins can read no active block and both insert one. The
     * customer row they each have to take first is what serialises them.
     * SQLite, which the app develops and tests on, serialises writers with a
     * database-level lock; its grammar compiles the clause away.
     *
     * @param  Builder<$this>  $query
     */
    #[Scope]
    protected function lockedForModeration(Builder $query): void
    {
        $query->lockForUpdate();
    }

    /**
     * Re-reads this row for update inside the caller's transaction, the way
     * `Fulfillment::takeForTransition` re-reads the row a transition is
     * judged against.
     */
    public function takeForModeration(): static
    {
        /** @var static $locked */
        $locked = $this->newQuery()->whereKey($this->getKey())->lockedForModeration()->sole();

        return $this->setRawAttributes($locked->getAttributes(), sync: true);
    }

    /**
     * The admin customers list, narrowed to one standing. `All` adds no
     * clause at all, which is what an empty filter asks for.
     *
     * @param  Builder<$this>  $query
     */
    #[Scope]
    protected function inStanding(Builder $query, StandingFilter $standing): void
    {
        match ($standing) {
            StandingFilter::All => $query,
            StandingFilter::Verified => $query->whereNotNull('email_verified_at'),
            StandingFilter::Anonymous => $query->whereNull('email'),
            StandingFilter::Blocked => $query->whereHas('blocks', fn (Builder $blocks): Builder => $blocks->whereNull('lifted_at')),
        };
    }

    /**
     * How the other side of a thread names this customer: their given name,
     * or their id. The address they verified stays with the admin console's
     * `displayName()`.
     */
    public function participantName(): string
    {
        return ParticipantName::forCustomer($this->name, $this->id);
    }
}
